File: client-js/kit/src/protocol/pending.js
// Política de retry pura sobre o mapa de pendentes (at-least-once).
// pending: Map ref -> {line, tick, retries}. Nada aqui toca no socket:
// onTimeout devolve o que fazer ({retry} ou {drop}) e a borda executa.
// Os mapas são tratados como imutáveis: cada função devolve um Map novo.

export const BUDGET = 3; // tentativas máximas por ref
export const TIMEOUT = 2; // ticks sem ACK antes de reenviar

// Registra uma ação enviada esperando ACK.
//   pending: Map ref -> {line, tick, retries}
//   ref: ref único da ação (ex: "mina-a1b2-3")
//   line: a linha de protocolo enviada (guardada para um possível retry)
//   tick: tick em que a ação foi enviada
// Devolve um NOVO mapa com a entrada adicionada (retries começa em 0).
// Exemplo:
//   addPending(new Map(), "r-1", "ACT r-1 expand 3 4", 10)
//   // => Map { "r-1" => { line: "ACT r-1 expand 3 4", tick: 10, retries: 0 } }
export function addPending(pending, ref, line, tick) {
  const next = new Map(pending);
  next.set(ref, { line, tick, retries: 0 });
  return next;
}

// Confirma um ref: o ACK chegou, a ação não precisa mais de retry.
//   pending: Map ref -> {line, tick, retries}
//   ref: ref confirmado pelo servidor
// Devolve um NOVO mapa sem o ref. Ref desconhecido é ignorado.
// Exemplo:
//   onAck(pending, "r-1")  // => mapa igual a pending, mas sem "r-1"
export function onAck(pending, ref) {
  const next = new Map(pending);
  next.delete(ref);
  return next;
}

// Vence os pendentes antigos no tick now. Para cada ref com
// now - tick >= timeout:
//   - se retries < budget: reenvia -> efeito { retry: ref, line },
//     retries+1 e tick atualizado para now
//   - se o budget esgotou: desiste -> efeito { drop: ref }, sai do mapa
// Refs recentes ficam intactos.
//   pending: Map ref -> {line, tick, retries}
//   now: tick atual
//   budget/timeout: limites (padrão BUDGET e TIMEOUT acima)
// Devolve [novoMapa, efeitos], na ordem de inserção dos refs.
// Exemplo:
//   const [next, effects] = onTimeout(pending, 12);
//   for (const eff of effects) {
//     if (eff.retry) conn.sendLine(eff.line); // reenvia a linha original
//     else log(`acao descartada sem ACK: ${eff.drop}`);
//   }
export function onTimeout(pending, now, budget = BUDGET, timeout = TIMEOUT) {
  const effects = [];
  const next = new Map();
  for (const [ref, entry] of pending) {
    if (now - entry.tick < timeout) {
      next.set(ref, entry);
    } else if (entry.retries < budget) {
      const retried = { ...entry, tick: now, retries: entry.retries + 1 };
      next.set(ref, retried);
      effects.push({ retry: ref, line: entry.line });
    } else {
      effects.push({ drop: ref });
    }
  }
  return [next, effects];
}

File: client-go/kit/pending.go
package main

// Política de retry pura sobre o mapa de pendentes.
// Nada aqui toca no socket: OnTimeout devolve o que fazer
// (retry ou drop) e a borda executa.
//
// INFRA: não edite este arquivo.

const (
	Budget  = 3 // tentativas máximas por ref
	Timeout = 2 // ticks sem ACK antes de reenviar
)

type PendingEntry struct {
	Line    string
	Tick    int
	Retries int
}

type Effect struct {
	Ref  string
	Line string // preenchida em retry
	Drop bool
}

// AddPending registra uma linha enviada que ainda espera ACK.
//
// Parâmetros:
//   - pending: mapa ref -> entrada pendente (tratado como imutável)
//   - ref: ref da linha enviada
//   - line: a linha completa enviada ao servidor (guardada para reenvio)
//   - tick: tick em que a linha foi enviada
//
// Devolve um NOVO mapa com a entrada adicionada (Retries começa em 0).
//
// Exemplo:
//
//	pending = AddPending(pending, ref, line, msg.Tick)
func AddPending(pending map[string]PendingEntry, ref, line string, tick int) map[string]PendingEntry {
	next := make(map[string]PendingEntry, len(pending)+1)
	for k, v := range pending {
		next[k] = v
	}
	next[ref] = PendingEntry{Line: line, Tick: tick}
	return next
}

// OnAck confirma um pendente: o ACK chegou, o ref sai do mapa.
//
// Parâmetros:
//   - pending: mapa ref -> entrada pendente
//   - ref: ref confirmado pelo servidor
//
// Devolve um NOVO mapa sem o ref confirmado.
//
// Exemplo:
//
//	pending = OnAck(pending, msg.Ref)
func OnAck(pending map[string]PendingEntry, ref string) map[string]PendingEntry {
	next := make(map[string]PendingEntry, len(pending))
	for k, v := range pending {
		if k != ref {
			next[k] = v
		}
	}
	return next
}

// OnTimeout vence os pendentes antigos e decide o que fazer com cada um.
//
// Para cada ref com now - Tick >= Timeout:
//   - se Retries < Budget: agenda reenvio (Effect com Line preenchida),
//     incrementa Retries e atualiza Tick para now
//   - se Retries esgotou Budget: desiste (Effect com Drop: true),
//     o ref some do mapa
//
// Refs recentes ficam intactos. Devolve o NOVO mapa e a lista de efeitos;
// quem chama executa os efeitos (reenvia a linha ou descarta o ref).
//
// Exemplo:
//
//	pending, effects = OnTimeout(pending, msg.Tick)
//	for _, e := range effects {
//		if !e.Drop {
//			send(e.Line)
//		}
//	}
func OnTimeout(pending map[string]PendingEntry, now int) (map[string]PendingEntry, []Effect) {
	var effects []Effect
	next := make(map[string]PendingEntry, len(pending))
	for ref, entry := range pending {
		switch {
		case now-entry.Tick < Timeout:
			next[ref] = entry
		case entry.Retries < Budget:
			entry.Tick = now
			entry.Retries++
			next[ref] = entry
			effects = append(effects, Effect{Ref: ref, Line: entry.Line})
		default:
			effects = append(effects, Effect{Ref: ref, Drop: true})
		}
	}
	return next, effects
}
